File: frontend/mobile_app/lib/data/journey_logger.dart
import 'dart:async';
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../core/config/app_config.dart';

/// Fire-and-forget salesperson clickstream logger (backend mode only). Records
/// controller-side actions that don't pass through the WS relay — add-to-cart,
/// opening recommendations, etc. — so the CMS activity log is a complete journey.
class JourneyLogger {
  JourneyLogger({http.Client? client}) : _client = client ?? http.Client();

  final http.Client _client;

  void log({
    required String eventType,
    required String? token,
    String? sessionId,
    String? refId,
    Map<String, dynamic>? meta,
  }) {
    if (!AppConfig.backendMode || token == null || token.isEmpty) return;
    unawaited(
      _client
          .post(
            AppConfig.http('/api/journey'),
            headers: <String, String>{
              'content-type': 'application/json',
              'authorization': 'Bearer $token',
            },
            body: jsonEncode(<String, dynamic>{
              'eventType': eventType,
              'sessionId': ?sessionId,
              'refId': ?refId,
              'meta': ?meta,
            }),
          )
          .timeout(const Duration(seconds: 5))
          .catchError((Object _) => http.Response('', 599)), // best-effort
    );
  }
}

File: frontend-rn/mobile_app/src/data/journeyLogger.ts
import { AppConfig } from '../config/appConfig';

/// Fire-and-forget salesperson clickstream logger (backend mode only). Records
/// controller-side actions that don't pass through the WS relay so the CMS
/// activity log is a complete journey. Ported 1:1 from `JourneyLogger`.
export class JourneyLogger {
  log(opts: {
    eventType: string;
    token?: string | null;
    sessionId?: string;
    refId?: string;
    meta?: Record<string, any>;
  }): void {
    if (!AppConfig.backendMode || opts.token == null || opts.token.length === 0) return;
    const body: Record<string, any> = { eventType: opts.eventType };
    if (opts.sessionId != null) body.sessionId = opts.sessionId;
    if (opts.refId != null) body.refId = opts.refId;
    if (opts.meta != null) body.meta = opts.meta;

    const controller = new AbortController();
    const t = setTimeout(() => controller.abort(), 5000);
    void fetch(AppConfig.http('/api/journey'), {
      method: 'POST',
      headers: { 'content-type': 'application/json', authorization: `Bearer ${opts.token}` },
      body: JSON.stringify(body),
      signal: controller.signal,
    })
      .catch(() => undefined)
      .finally(() => clearTimeout(t));
  }
}
